fetchurl: set SSL_CERT_FILE based on finalAttrs

Scrumplex opened the pull request:
This can cause an insecure derivation when overriding deriviations with empty output hashes (as done by nix-update)

See https://github.com/Mic92/nix-update/issues/397


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Scrumplex committed:
fetchurl: set SSL_CERT_FILE based on finalAttrs

This can cause an insecure derivation when overriding deriviations with
empty output hashes (as done by nix-update)

See https://github.com/Mic92/nix-update/issues/397

Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/idea-nix-fod-or-whatever-other-better-name/66210/16

Scrumplex commented:
eval results:

# Performance comparison

This compares the performance of this branch against its pull request base branch (e.g., 'master')

For further help please refer to: [ci/README.md](https://github.com/NixOS/nixpkgs/blob/master/ci/README.md)


| metric | mean_before | mean_after | mean_diff | mean_%_change | p_value | t_stat |
| - | - | - | - | - | - | - |
| cpuTime | 34.6244 | 35.5173 | 0.8929 | 2.7654 | 0.0000 | 6.0484 |
| envs.bytes | 722236245.7143 | 722238247.6190 | 2001.9048 | 0.0002 | 0.0262 | 2.3061 |
| envs.elements | 52532706.7143 | 52532831.8333 | 125.1190 | 0.0001 | 0.0262 | 2.3061 |
| envs.number | 37746824.0000 | 37746949.1190 | 125.1190 | 0.0002 | 0.0262 | 2.3061 |
| gc.cycles | 10.0238 | 10.0238 | - | - | - | - |
| gc.heapSize | 2919987590.0952 | 2930773040.7619 | 10785450.6667 | 0.4331 | 0.0201 | 2.4187 |
| gc.totalBytes | 5428494059.8095 | 5446079020.9524 | 17584961.1429 | 0.3206 | 0.0000 | 11.3671 |
| list.bytes | 101814500.1905 | 101814500.1905 | - | - | - | - |
| list.concats | 2267136.1905 | 2267136.1905 | - | - | - | - |
| list.elements | 12726812.5238 | 12726812.5238 | - | - | - | - |
| nrAvoided | 41709734.8095 | 41773615.9762 | 63881.1667 | 0.1510 | 0.0000 | 10.8709 |
| nrExprs | 1751179.6429 | 1751180.6429 | 1.0000 | 0.0001 | - | inf |
| nrFunctionCalls | 34583653.9286 | 34583653.9286 | - | - | - | - |
| nrLookups | 15902280.3333 | 15902405.4524 | 125.1190 | 0.0005 | 0.0262 | 2.3061 |
| nrOpUpdateValuesCopied | 88463730.5000 | 89213917.9048 | 750187.4048 | 0.8395 | 0.0000 | 11.4927 |
| nrOpUpdates | 3667181.5476 | 3698559.0714 | 31377.5238 | 0.8321 | 0.0000 | 11.3811 |
| nrPrimOpCalls | 18657450.0476 | 18657450.0476 | - | - | - | - |
| nrThunks | 44221401.9762 | 44222778.2857 | 1376.3095 | 0.0019 | 0.0262 | 2.3061 |
| sets.bytes | 1975007129.9048 | 1989058200.7619 | 14051070.8571 | 0.7095 | 0.0000 | 11.3930 |
| sets.elements | 116506831.5238 | 117322151.2857 | 815319.7619 | 0.6969 | 0.0000 | 11.3986 |
| sets.number | 6931114.0952 | 6993986.2619 | 62872.1667 | 0.9495 | 0.0000 | 11.3210 |
| sizes.Attr | 16.0000 | 16.0000 | - | - | - | - |
| sizes.Bindings | 16.0000 | 16.0000 | - | - | - | - |
| sizes.Env | 8.0000 | 8.0000 | - | - | - | - |
| sizes.Value | 24.0000 | 24.0000 | - | - | - | - |
| symbols.bytes | 1526738.8095 | 1526738.8095 | - | - | - | - |
| symbols.number | 118400.5476 | 118400.5476 | - | - | - | - |
| time.cpu | 34.6244 | 35.5173 | 0.8929 | 2.7654 | 0.0000 | 6.0484 |
| time.gc | 5.7797 | 5.9958 | 0.2161 | 3.9642 | 0.0004 | 3.8551 |
| time.gcFraction | 0.1929 | 0.1945 | 0.0016 | 1.1349 | 0.1975 | 1.3101 |
| values.bytes | 1444959578.2857 | 1444992609.7143 | 33031.4286 | 0.0014 | 0.0262 | 2.3061 |
| values.number | 60206649.0952 | 60208025.4048 | 1376.3095 | 0.0014 | 0.0262 | 2.3061 |

# Packages

<details> <summary>Added packages (0)</summary>

</details>

<details> <summary>Removed packages (0)</summary>

</details>

<details> <summary>Changed packages (0)</summary>

</details>

tejing1 commented:
Should this instead be converting `overrideAttrs` on the fetcher to override the fetcher args rather than the `mkDerivation` args? Much as we've already done for most of the ecosystem-specific build functions?

tejing1 reviewed (commented):
Once you sort through the formatting changes, it looks good to me from reading the code. Haven't tested locally.

Pandapip1 commented:
I'd say it probably makes sense to fix the security issue first, and *then* worry about eval performance in a follow-up PR.

tejing1 commented:
Regardless of other changes, I'd put a bit of shell code in the derivation itself that checks at fetch time if `outputHash` and `SSL_CERT_FILE` form a dangerous combination:

`builder.sh`:
```bash
if [ -z "$allowUnsafePrefetch" ] && [ "$SSL_CERT_FILE" == "/no-cert-file.crt" ]; then
  for prefetchHash in $prefetchHashes; do
    [ "$outputHash" == "$prefetchHash" ] && echo "error: Refusing non-SSL-protected prefetch" && exit 1
  done
fi
```
`default.nix`:
```nix
prefetchHashes = [ "" lib.fakeHash lib.fakeSha256 lib.fakeSha512 ];
```

Scrumplex committed:
fetchurl: harden SSL_CERT_FILE logic in builder

This should serve as a last line of defense, in case of a regression in
the fetchurl derivation.

See https://github.com/NixOS/nixpkgs/pull/420978#issuecomment-3020393211

Co-authored-by: Jeff Huffman <tejing@tejing.com>
Signed-off-by: Sefa Eyeoglu <contact@scrumplex.net>

tejing1 commented on pkgs/build-support/fetchurl/builder.sh:
```diff
@@ -19,6 +19,12 @@ curl=(
 )
 
 if ! [ -f "$SSL_CERT_FILE" ]; then
+    for prefetchHash in $prefetchHashes; do
```
```bash
exec {fd}< <(printf "%s" "$prefetchHashes")
readarray -u $fd -d " " -t prefetchHashes
exec {fd}<&-
for prefetchHash in "${prefetchHashes[@]}" do
  if [ "$outputHash" == "$prefetchHash" ]; then
    echo "error: Refusing non-SSL-protected prefetch."
    exit 1
  fi
  curl+=(--insecure)
done
```
Should work properly without being *too* unreadable. I feel like there should be a more straightforward way, but I'm not seeing it.
